### PR TITLE
Change Linting engine from `TSLint` to `ESLint`

### PR DESCRIPTION
**Pull request by mhenkens**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/NationalBankBelgium/stark/blob/master/CONTRIBUTING.md#-commit-message-guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[x] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Using ESLint insteadof TSLint

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

Somme manual adaptation must be done before. An angular schematics will be provided to help the migration

## Other information

### CODE REVIEW

File: packages/stark-ui/.eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
+{
+  "extends": "../../.eslintrc.json",
+  "ignorePatterns": ["!**/*"],
+  "overrides": [
+    {
```
Comment (SuperITMan) [inline packages/stark-ui/.eslintrc.json:5]: This file could be simplified since the rules are already defined in the root file:

```json
{
  "extends": "../../.eslintrc.json",
  "ignorePatterns": ["!**/*"],
  "overrides": [
    {
      "files": ["*.spec.ts"],
      "rules": {
        "@angular-eslint/component-selector": "off",
        "@angular-eslint/directive-selector": "off",
        "@typescript-eslint/unbound-method": "off"
      }
    }
  ]
}
```

File: packages/stark-ui/.eslintrc.json [hunk 2]
```diff
@@ -0,0 +1,41 @@
+{
```
Comment (SuperITMan) [inline packages/stark-ui/.eslintrc.json:1]: This file should be included in the stark-ui commit instead of stark-rbac

File: packages/stark-ui/tslint.json
Comment (SuperITMan) [inline packages/stark-ui/tslint.json:1]: This file removal should be part of the stark-ui commit instead of stark-rbac commit

File: starter/e2e/tsconfig.app.json
Comment (SuperITMan) [inline starter/e2e/tsconfig.app.json:1]: Is this change related to ESLint ?

File: angular.json [hunk 3]
```diff
@@ -85,10 +84,9 @@
           }
         },
         "lint": {
-          "builder": "@angular-devkit/build-angular:tslint",
+          "builder": "@angular-eslint/builder:lint",
```
Comment (SuperITMan) [inline angular.json:87]: Maybe this change could be part of the stark-ui commit ?

File: showcase/.eslintrc.json [hunk 4]
```diff
@@ -0,0 +1,38 @@
+{
+  "root": true,
```
Comment (SuperITMan) [inline showcase/.eslintrc.json:2]: This file could be simplified since the rules are already defined in the eslint-config:

```json
{
  "root": true,
  "extends": "@nationalbankbelgium",
  "ignorePatterns": ["src/assets/**/*"],
  "overrides": [
    {
      "files": ["*.ts"],
      "parserOptions": {
        "project": ["tsconfig.json", "e2e/tsconfig.e2e.json"],
        "createDefaultProgram": true
      },
      "rules": {
        "@typescript-eslint/no-unused-vars": ["error", { "argsIgnorePattern": "^_" }],
        "@angular-eslint/component-selector": [
          "error",
          {
            "prefix": ["app", "demo", "showcase", "stark"],
            "style": "kebab-case",
            "type": "element"
          }
        ],
        "@angular-eslint/directive-selector": [
          "error",
          {
            "prefix": ["app", "demo", "showcase"],
            "style": "camelCase",
            "type": "attribute"
          }
        ]
      }
    }
  ]
}
```

File: showcase/angular.json [hunk 5]
```diff
@@ -195,5 +194,8 @@
     "@schematics/angular:directive": {
       "prefix": "showcase"
     }
+  },
+  "cli": {
```
Comment (SuperITMan) [inline showcase/angular.json:198]: Is this configuration required ?

File: showcase/e2e/.eslintrc.json [hunk 6]
```diff
@@ -0,0 +1,35 @@
+{
```
Comment (SuperITMan) [inline showcase/e2e/.eslintrc.json:1]: This file could be simplified since the rules are already defined in the root file:

```json
{
  "extends": "../.eslintrc.json",
  "ignorePatterns": ["!**/*"],
  "overrides": [
    {
      "files": ["*.ts"],
      "parserOptions": {
        "project": ["tsconfig.e2e.json"],
        "createDefaultProgram": true
      },
      "rules": {
        "@angular-eslint/directive-selector": [
          "error",
          {
            "type": "attribute",
            "prefix": "",
            "style": "camelCase"
          }
        ],
        "@angular-eslint/component-selector": [
          "error",
          {
            "type": "element",
            "prefix": "",
            "style": "kebab-case"
          }
        ]
      }
    }
  ]
}
```

File: showcase/package.json [hunk 7]
```diff
@@ -26,37 +26,38 @@
...
-    "github-deploy": "npm run github-deploy:dev",
     "karma": "karma",
-    "lint": "npm run lint-ts && npm run lint-css",
-    "lint-ts": "tslint --config ./tslint.json --project ./tsconfig.json --format codeFrame",
-    "lint-ts:fix": "node --max_old_space_size=4096 ./node_modules/tslint/bin/tslint --config ./tslint.json --project ./tsconfig.json --format codeFrame --fix",
+    "lint": "npm run lint-es && npm run lint-css",
     "lint-css": "stylelint \"./src/**/*.?(pc|sc|c|sa)ss\" --formatter \"string\"",
+    "lint-es": "ng lint",
+    "lint-es:fix": "ng lint --fix",
+    "lint-ts:fix": "node --max_old_space_size=4096 ./node_modules/tslint/bin/tslint --config ./tslint.json --project ./tsconfig.json --format codeFrame --fix",
```
Comment (SuperITMan) [inline showcase/package.json:60]: Could you remove this "lint-ts:fix" script since it is no longer usable?

File: showcase/src/app/app.component.ts [hunk 8]
```diff
@@ -31,7 +31,7 @@ import { APP_MENU_CONFIG } from "./app-menu.config";
  * Top Level Component
  */
 @Component({
-	selector: "app",
+	selector: "app-showcase",
```
Comment (SuperITMan) [inline showcase/src/app/app.component.ts:34]: Is this change required ?

File: showcase/src/app/shared/components/example-viewer/example-viewer.component.ts [hunk 9]
```diff
@@ -71,8 +72,8 @@ export class ExampleViewerComponent implements OnInit {
 	 */
 	public getAppBaseHref(): string {
 		// the final url in GitHub Pages will be something like "/showcase/latest/" or "/showcase/some-version/"
-		const finalUrlRegex: RegExp = /(\/showcase\/[\d\D][^\/]+(\/|\/$|$))/;
-		const trailingSlashRegex: RegExp = /\/$/;
+		const finalUrlRegex = /(\/showcase\/[\d\D][^/]+(\/|\/$|$))/;
```
Comment (SuperITMan) [inline showcase/src/app/shared/components/example-viewer/example-viewer.component.ts:75]: Why the character `\` before the `/` has been removed ?

File: showcase/src/app/welcome/pages/reactive-form-errors/components/translated-form-error/translated-form-error.component.ts [hunk 10]
```diff
@@ -18,7 +18,7 @@ export class TranslatedFormErrorComponent implements NgxFormErrorComponent, OnIn
 	public constructor(public translateService: TranslateService) {}
 
 	public ngOnInit(): void {
-		this.translateService.onLangChange.subscribe((_ev: LangChangeEvent) => {
+		this.translateService.onLangChange.subscribe((_: LangChangeEvent) => {
```
Comment (SuperITMan) [inline showcase/src/app/welcome/pages/reactive-form-errors/components/translated-form-error/translated-form-error.component.ts:21]: Why changing the variable name?

File: starter/.eslintrc.json [hunk 11]
```diff
@@ -0,0 +1,37 @@
+{
+  "root": true,
```
Comment (SuperITMan) [inline starter/.eslintrc.json:2]: This file could be simplified since the rules are already defined in the eslint-config:

```json
{
  "root": true,
  "extends": ["@nationalbankbelgium"],
  "overrides": [
    {
      "files": ["*.ts"],
      "parserOptions": {
        "project": ["tsconfig.json", "e2e/tsconfig.json"],
        "createDefaultProgram": true
      },
      "rules": {
        "@angular-eslint/component-selector": [
          "error",
          {
            "prefix": "app",
            "style": "kebab-case",
            "type": "element"
          }
        ],
        "@angular-eslint/directive-selector": [
          "error",
          {
            "prefix": "app",
            "style": "camelCase",
            "type": "attribute"
          }
        ]
      }
    }
  ]
}
```

File: starter/angular.json [hunk 12]
```diff
@@ -193,5 +191,8 @@
     "@schematics/angular:directive": {
       "prefix": "app"
     }
+  },
+  "cli": {
```
Comment (SuperITMan) [inline starter/angular.json:195]: Is this configuration required?

File: starter/e2e/.eslintrc.json [hunk 13]
```diff
@@ -0,0 +1,35 @@
+{
+  "extends": "../../.eslintrc.json",
```
Comment (SuperITMan) [inline starter/e2e/.eslintrc.json:2]: This file could be simplified since the rules are already defined in the root file:

```json
{
  "extends": "../.eslintrc.json",
  "ignorePatterns": ["!**/*"],
  "overrides": [
    {
      "files": ["*.ts"],
      "parserOptions": {
        "project": ["tsconfig.e2e.json"],
        "createDefaultProgram": true
      },
      "rules": {
        "@angular-eslint/directive-selector": [
          "error",
          {
            "type": "attribute",
            "prefix": "",
            "style": "camelCase"
          }
        ],
        "@angular-eslint/component-selector": [
          "error",
          {
            "type": "element",
            "prefix": "",
            "style": "kebab-case"
          }
        ]
      }
    }
  ]
}
```

File: package-lock.json
Comment (SuperITMan) [inline package-lock.json:1]: Could you move this change in the first commit, where you change the dependencies in the package-lock.json ?

**Review by SuperITMan — CHANGES_REQUESTED** [review]
A few remarks before we can merge this into the `master` branch 👍 

Could you search for all `/* eslint-disable */` occurrences and adapt them to disable only the problematic rule ?

File: showcase/src/app/app.component.ts (hunk 8 above)
Comment (mhenkens) [reply showcase/src/app/app.component.ts:34]: the prefix must be followed by something.

File: showcase/src/app/shared/components/example-viewer/example-viewer.component.ts (hunk 9 above)
Comment (mhenkens) [reply showcase/src/app/shared/components/example-viewer/example-viewer.component.ts:75]: those character are not  necessary and break the rule. This is done automatically with the --fix option

**Review by SuperITMan — CHANGES_REQUESTED** [review]
Still a few changes to apply, we are close to merge this 👍

File: package.json [hunk 14]
```diff
@@ -223,13 +223,6 @@
     "test:ci:starter": "cd starter && npm run test-fast:ci && cd ../..",
     "test:ci:coveralls:combined": "node combine-packages-coverage.js",
     "tsc": "tsc",
-    "tslint": "tslint",
```
[the remark this answers is not in the document]
Comment (mhenkens) [reply package.json:226]: Added

File: packages/stark-build/config/build-utils.js [hunk 15]
```diff
@@ -112,45 +112,19 @@ function get(obj, path, defaultValue) {
...
- *
- * @param {string} tslintConfig The path to the projects tslint.json
- * @return {string} Returns the path to the temporary tslint configuration for tslint-loader
- */
-function getFixedTSLintConfig(tslintConfig) {
-	// Windows style separators need to be replaced before injecting it into the new tslint.json
-	const tslintConfigPath = helpers.root(tslintConfig).replace(/\\/g, "/");
-
-	const noTypeCheckTSLintConfig = fs
-		.readFileSync(helpers.rootStark("config/tslint-disabled-typecheck-rules.json"), "utf8")
```
[the remark this answers is not in the document]
Comment (mhenkens) [reply packages/stark-build/config/build-utils.js:127]: removed

File: showcase/src/app/demo-ui/pages/generic-search/services/demo-generic.service.ts [hunk 16]
```diff
@@ -129,7 +129,7 @@ export class DemoGenericService implements StarkGenericSearchService<HeroMovie,
 			}
 		}
 
-		// tslint:disable-next-line:no-alphabetical-sort
+		// eslint-disable-next-line
```
[the remark this answers is not in the document]
Comment (mhenkens) [reply showcase/src/app/demo-ui/pages/generic-search/services/demo-generic.service.ts:132]: removed

File: packages/stark-ui/.eslintrc.json [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+  "extends": "../../.eslintrc.json",
+  "ignorePatterns": ["!**/*"],
+  "overrides": [
+    {
+      "files": ["*.ts"],
+      "excludedFiles": ["*.spec.ts"],
```
[the remark this answers is not in the document]
Comment (mhenkens) [reply packages/stark-ui/.eslintrc.json:7]: Because we override rules for the files : ["*.ts"] all file ["*.ts"] are scan. those rule must be override because the prefix for component and directive id defined in this file. to avoid error in test where we create dummy components we should add "excludedFiles": ["*.spec.ts"]

File: packages/stark-core/.eslintrc.json [hunk 18]
```diff
@@ -0,0 +1,32 @@
+{
+  "extends": "../../.eslintrc.json",
+  "ignorePatterns": ["!**/*"],
+  "overrides": [
+    {
+      "files": ["*.ts"],
+      "excludedFiles": ["*.spec.ts"],
```
[the remark this answers is not in the document]
Comment (mhenkens) [reply packages/stark-core/.eslintrc.json:7]: Because we override rules for the files : [".ts"] all file [".ts"] are scan. those rule must be override because the prefix for component and directive id defined in this file. to avoid error in test where we create dummy components we should add "excludedFiles": ["*.spec.ts"]

File: packages/stark-core/src/modules/xsrf/services/xsrf.service.ts [hunk 19]
```diff
@@ -221,7 +220,7 @@ export class StarkXSRFServiceImpl implements StarkXSRFService {
 	}
 
 	private getCookieRegExp(cookieName: string): RegExp {
-		const escapedName: string = cookieName.replace(/([\[\]\{\}\(\)\|\=\;\+\?\,\.\*\^\$])/gi, "\\$1");
+		const escapedName: string = cookieName.replace(/([[]{}()|=;+?,.*^$])/gi, "\\$1");
```
[the remark this answers is not in the document]
Comment (mhenkens) [reply packages/stark-core/src/modules/xsrf/services/xsrf.service.ts:223]: The rule https://eslint.org/docs/latest/rules/no-useless-escape ask to remove the useless escape char and both regexp behave the same.

File: packages/stark-ui/src/modules/dropdown/components/dropdown.component.ts [hunk 20]
```diff
@@ -549,10 +553,10 @@ export class StarkDropdownComponent
 	 */
 	public onContainerClick(): void {
 		// Mimic implementation of MatSelect: https://github.com/angular/components/blob/master/src/material/select/select.ts
-		if (!!this.singleSelectElement) {
+		if (this.singleSelectElement) {
```
[the remark this answers is not in the document]
Comment (mhenkens) [reply packages/stark-ui/src/modules/dropdown/components/dropdown.component.ts:556]: because this raise the linting error no-extra-boolean-cast
see https://eslint.org/docs/latest/rules/no-extra-boolean-cast

File: packages/stark-ui/src/modules/dropdown/components/dropdown.component.ts [hunk 21]
```diff
@@ -549,10 +553,10 @@ export class StarkDropdownComponent
...
 			this.singleSelectElement.focus();
 			this.singleSelectElement.open();
-		} else if (!!this.multiSelectElement) {
+		} else if (this.multiSelectElement) {
```
[the remark this answers is not in the document]
Comment (mhenkens) [reply packages/stark-ui/src/modules/dropdown/components/dropdown.component.ts:559]: because this raise the linting error no-extra-boolean-cast
see https://eslint.org/docs/latest/rules/no-extra-boolean-cast

File: packages/stark-core/src/modules/routing/services/routing.service.ts [hunk 22]
```diff
@@ -332,15 +331,15 @@ export class StarkRoutingServiceImpl implements StarkRoutingService {
...
-								currentState: <string>fromState.name,
-								newState: targetState.name().toString(),
-								params: targetState.params(),
-								error: rejectionString
-							});
-
 							// dispatch corresponding action to allow the user to trigger his own effects if needed
-							this.store.dispatch(failureAction);
+							this.store.dispatch(
+								StarkRoutingActions.navigateFailure({
```
[the remark this answers is not in the document]
Comment (mhenkens) [reply packages/stark-core/src/modules/routing/services/routing.service.ts:336]: This raise the error no-case-declarations https://eslint.org/docs/latest/rules/no-case-declarations

File: .eslintrc.json [hunk 23]
```diff
@@ -0,0 +1,18 @@
+{
+  "root": true,
+  "ignorePatterns": ["projects/**/*"],
```
[the remark this answers is not in the document]
Comment (SuperITMan) [reply .eslintrc.json:3]: There is no "projects" folder in Stark. Could you explain why do we need this `ignorePatterns` ?

File: packages/stark-ui/src/modules/session-ui/routes.ts [hunk 24]
```diff
@@ -40,7 +41,7 @@ export function destroyOverlaysOnEnterFn(transition: Transition): HookResult {
 /**
  * States defined by Session-UI Module
  */
-/* tslint:disable:no-duplicate-string */
+
```
[the remark this answers is not in the document]
Comment (SuperITMan) [reply packages/stark-ui/src/modules/session-ui/routes.ts:44]: Could you have a look at this ?

File: starter/package.json [hunk 25]
```diff
@@ -134,14 +134,24 @@
     "zone.js": "~0.11.4"
   },
   "devDependencies": {
+    "@angular-eslint/builder": "12.7.0",
+    "@angular-eslint/eslint-plugin": "12.7.0",
+    "@angular-eslint/eslint-plugin-template": "12.7.0",
+    "@angular-eslint/schematics": "12.7.0",
+    "@angular-eslint/template-parser": "12.7.0",
```
[the remark this answers is not in the document]
Comment (SuperITMan) [reply starter/package.json:141]: Could you have a look at this ?

**Review by SuperITMan — CHANGES_REQUESTED** [review]
Still a few changes to apply, we're close to being able to merge this 👍

File: showcase/.eslintrc.json [hunk 26]
```diff
@@ -0,0 +1,39 @@
...
+          "error",
+          {
+            "prefix": ["app", "demo", "showcase"],
+            "style": "camelCase",
+            "type": "attribute"
+          }
+        ]
+      }
+    },
+    {
```
[the remark this answers is not in the document]
Comment (mhenkens) [reply showcase/.eslintrc.json:33]: No is needed for the prefix for the component-selector and  directive-selector

File: starter/.eslintrc.json [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+        ]
+      }
+    },
+    {
+      "files": ["*.html"],
+      "extends": ["@nationalbankbelgium"],
+      "rules": {}
+    }
+  ]
+}
```
Comment (SuperITMan) [inline starter/.eslintrc.json:39]: Could you remove this override ?
As it is already extended in the root configuration at the top of the file.

```suggestion
        ]
      }
    }
  ]
}
```

**Review by SuperITMan — CHANGES_REQUESTED** [review]
3 small changes to apply then we can merge this PR 👍

**Review by SuperITMan — APPROVED** [review]
LGTM 👍
